blocks/bricks/bn.py

```diff
@@ -0,0 +1,345 @@
+        # learned using the same cost function as other model parameters.
+        self.population_mean = shared_floatx_zeros(var_dim,
+                                                   name='population_mean',
+                                                   broadcastable=broadcastable)
+        self.population_stdev = shared_floatx(numpy.ones(var_dim),
+                                              name='population_stdev',
+                                              broadcastable=broadcastable)
+        add_role(self.population_mean, BATCH_NORM_POPULATION_MEAN)
+        add_role(self.population_stdev, BATCH_NORM_POPULATION_STDEV)
+
+    @property
+    def W(self):
+        return self._W
+
+    @property
+    def b(self):
+        return self._b
+
+    def _initialize(self):
+        self.biases_init.initialize(self.b, self.rng)
+        self.weights_init.initialize(self.W, self.rng)
+
+    # Needed for the Feedforward interface.
+    @property
+    def output_dim(self):
+        return self.input_dim
+
+    # The following properties allow for BatchNormalization bricks
+    # to be used directly inside of a ConvolutionalSequence.
+    @property
+    def image_size(self):
+        return self.input_dim[-2:]
+
+    @image_size.setter
+    def image_size(self, value):
+        if not isinstance(self.input_dim, collections.Sequence):
+            self.input_dim = (None,) + tuple(value)
+        else:
+            self.input_dim = (self.input_dim[0],) + tuple(value)
+
+    @property
+    def num_channels(self):
+        return self.input_dim[0]
+
+    @num_channels.setter
+    def num_channels(self, value):
+        if not isinstance(self.input_dim, collections.Sequence):
+            self.input_dim = (value,) + (None, None)
+        else:
+            self.input_dim = (value,) + self.input_dim[-2:]
+
+    def get_dim(self, name):
+        if name in ('input', 'output'):
+            return self.input_dim
+        else:
+            raise KeyError
+
+    @property
+    def num_output_channels(self):
+        return self.num_channels
+
+
+class SpatialBatchNormalization(BatchNormalization):
+    """Convenient subclass for batch normalization across spatial inputs.
+
+    Parameters
+    ----------
+    input_dim : int or tuple
+        The input size of a single example. Must be length at least 2.
+        It's assumed that the first axis of this tuple is a "channels"
+        axis, which should not be summed over, and all remaining
+        dimensions are spatial dimensions.
+
+    Notes
+    -----
+    See :class:`BatchNormalization` for more details (and additional
+    keyword arguments).
+
+    """
+    @lazy(allocation=['input_dim'])
+    def __init__(self, input_dim, **kwargs):
+        if not isinstance(input_dim,
+                          collections.Sequence) or len(input_dim) < 2:
+            raise ValueError('expected input_dim to be length >= 2 '
+                             '(channels, height, width)')
```

Comment:
The condition is for at least 2 dimensions whereas the error lists 3. Requiring at least 2 dimensions seems right to me as it allows for 1D convolutions if we ever get those. Maybe put an "e.g." in front of the list of dimensions in the error message.